akka-http-core/src/test/scala/akka/http/engine/client/ConnectionPoolSpec.scala

```diff
@@ -0,0 +1,358 @@
+/*
+ * Copyright (C) 2009-2014 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package akka.http.engine.client
+
+import java.util.concurrent.atomic.AtomicInteger
+import scala.concurrent.Await
+import scala.concurrent.duration._
+import scala.util.{ Failure, Success, Try }
+import akka.http.engine.server.ServerSettings
+import akka.http.util.{ SingletonException, StreamUtils }
+import akka.util.ByteString
+import akka.stream.{ BidiShape, ActorFlowMaterializer }
+import akka.stream.testkit.{ AkkaSpec, StreamTestKit }
+import akka.stream.scaladsl._
+import akka.http.{ Http, TestUtils }
+import akka.http.model.headers._
+import akka.http.model._
+
+class ConnectionPoolSpec extends AkkaSpec("akka.loggers = []\n akka.loglevel = OFF\n akka.io.tcp.trace-logging = off") {
+  implicit val materializer = ActorFlowMaterializer()
+
+  "The host-level client infrastructure" should {
+
+    "properly complete a simple request/response cycle" in new TestSetup {
+      val (requestIn, responseOut, responseOutSub, hcp) = cachedHostConnectionPool[Int]()
+
+      val requestInSub = requestIn.expectSubscription()
+      requestInSub.expectRequest(4)
+      requestInSub.sendNext(HttpRequest(uri = "/") -> 42)
+
+      acceptIncomingConnection()
+      responseOutSub.request(1)
+      val (Success(response), 42) = responseOut.expectNext()
+      response.headers should contain(RawHeader("Req-Host", s"localhost:$serverPort"))
+    }
+
+    "add default headers to requests if they don't contain them" in new TestSetup {
+      val (requestIn, responseOut, responseOutSub, hcp) =
+        cachedHostConnectionPool[Int](defaultHeaders = List(RawHeader("X-Custom-Header", "Default")))
+
+      val requestInSub = requestIn.expectSubscription()
+      requestInSub.expectRequest(4)
+      requestInSub.sendNext(HttpRequest(uri = "/") -> 42)
+
+      acceptIncomingConnection()
+      responseOutSub.request(1)
+      val (Success(response), 42) = responseOut.expectNext()
+      response.headers should contain(RawHeader("Req-X-Custom-Header", "Default"))
+    }
+
+    "open a second connection if the first one is loaded" in new TestSetup {
+      val (requestIn, responseOut, responseOutSub, hcp) = cachedHostConnectionPool[Int]()
+
+      val requestInSub = requestIn.expectSubscription()
+      requestInSub.expectRequest(4)
+      requestInSub.sendNext(HttpRequest(uri = "/a") -> 42)
+      requestInSub.sendNext(HttpRequest(uri = "/b") -> 43)
+
+      responseOutSub.request(2)
+      acceptIncomingConnection()
```

Comment:
so it also opens a second connection while the first is being established? otherwise it would be more reliable to cork the server’s processing until the second connection is made